flatmap/expand.go

```diff
@@ -42,9 +43,39 @@ func expandArray(m map[string]string, prefix string) []interface{} {
 		panic(err)
 	}
 
+	keySet := map[int]bool{}
```

Comment:
Your PR explanation is great, but I'd also add a short comment above here to explain why we're doing this versus the normal incremental range. Changing it would of course fail a test, but in the future I don't want a maintainer to think "but why did jbardin do this?!" Yes, you could always `git blame`, which is why I said "short comment" to explain the gist and then say something like "see GH-11042 for full details"